// @nestjs/testing/testing-module.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const shared_utils_1 = require("@nestjs/common/utils/shared.utils");
const module_token_factory_1 = require("@nestjs/core/injector/module-token-factory");
const unknown_module_exception_1 = require("./errors/unknown-module.exception");
const core_1 = require("@nestjs/core");
class TestingModule {
    constructor(container, scope, contextModule) {
        this.container = container;
        this.scope = scope;
        this.contextModule = contextModule;
        this.moduleTokenFactory = new module_token_factory_1.ModuleTokenFactory();
    }
    createNestApplication(express) {
        return new core_1.NestApplication(this.container, express);
    }
    createNestMicroservice(config) {
        return new core_1.NestMicroservice(this.container, config);
    }
    select(module) {
        const modules = this.container.getModules();
        const moduleMetatype = this.contextModule.metatype;
        const scope = this.scope.concat(moduleMetatype);
        const token = this.moduleTokenFactory.create(module, scope);
        const selectedModule = modules.get(token);
        if (!selectedModule) {
            throw new unknown_module_exception_1.UnknownModuleException();
        }
        return new TestingModule(this.container, scope, selectedModule);
    }
    get(metatypeOrToken) {
        return this.findInstanceByPrototypeOrToken(metatypeOrToken);
    }
    findInstanceByPrototypeOrToken(metatypeOrToken) {
        const dependencies = new Map([
            ...this.contextModule.components,
            ...this.contextModule.routes,
            ...this.contextModule.injectables,
        ]);
        const name = shared_utils_1.isFunction(metatypeOrToken) ? metatypeOrToken.name : metatypeOrToken;
        const instanceWrapper = dependencies.get(name);
        if (!instanceWrapper) {
            return null;
        }
        return instanceWrapper.instance;
    }
}
exports.TestingModule = TestingModule;
