// @nestjs/websockets/adapters/io-adapter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const io = require("socket.io");
const constants_1 = require("./../constants");
const Observable_1 = require("rxjs/Observable");
require("rxjs/add/observable/fromEvent");
require("rxjs/add/operator/switchMap");
require("rxjs/add/operator/filter");
class IoAdapter {
    create(port) {
        return io(port);
    }
    createWithNamespace(port, namespace) {
        return io(port).of(namespace);
    }
    bindClientConnect(server, callback) {
        server.on(constants_1.CONNECTION_EVENT, callback);
    }
    bindClientDisconnect(client, callback) {
        client.on(constants_1.DISCONNECT_EVENT, callback);
    }
    bindMessageHandlers(client, handlers, process) {
        handlers.forEach(({ message, callback }) => Observable_1.Observable.fromEvent(client, message)
            .switchMap((data) => process(callback(data)))
            .filter((result) => !!result && result.event)
            .subscribe(({ event, data }) => client.emit(event, data)));
    }
    bindMiddleware(server, middleware) {
        server.use(middleware);
    }
}
exports.IoAdapter = IoAdapter;
